rasterio/_io.pyx

```diff
@@ -820,13 +823,168 @@ cdef class DatasetReaderBase(DatasetBase):
         return sample_gen(self, xy, indexes)
 
 
+cdef class MemoryFileBase(object):
+    """Base for a BytesIO-like class backed by an in-memory file."""
+
```

Comment:
I've had the same thought, but there's a big difference between `MemoryFile` and S3 or HTTPs resources. The latter exist outside our Python processes. We can make them or modify them using APIs, yes, but external servers decide their state. A `MemoryFile` exists inside our Python process and not only do we decide its state, we need to deliberately keep it open, because closing a `MemoryFile` unlinks the in-memory file. We need a little extra abstraction in this case, and `MemoryFile` is it.